documentation/incremental_settings.md

```diff
@@ -0,0 +1,76 @@
+# Incremental Settings for Ansible Values
+
+The general approach in Ansible is that you provide all of the values for the
+current state of the managed object.  But this doesn't work when you have to
+manage the state of hundreds of values.  For example, when you use the
+`sysctl` module, you do not provide every single `sysctl` setting describing
+the entire state of the kernel settings.  Instead, you provide a single
+setting, which *incrementally* updates the state of all of the `sysctl`
+settings, by adding, modifying, and removing single values.
+```yaml
+- name: modify a single kernel setting
+  sysctl:
+    name: some_setting
+    value: 123  # to add/replace the value
+```
+or
+```yaml
+- name: modify a single kernel setting
+  sysctl:
+    name: some_setting
+    state: absent  # to remove the value
+```
+In the System Roles project, we had a need to generalize this for lists of
+settings and sections of settings.  We based our approach on
+[Kubernetes JSON strategic merge](https://stupefied-goodall-e282f7.netlify.app/contributors/devel/strategic-merge-patch/),
+and reuse of the Ansible `state` keyword.  We use this to manage objects which
+are represented by a `list` of `dict` objects.
+
+## Removing list items with state: absent
+
+This is the case where you want to remove some of the values from the managed
+object. Use `state: absent` in the `dict` value to remove the item named by
+`name`. For example, with `kernel_settings_sysctl` parameters:
+```yaml
+kernel_settings_sysctl:
+  - name: fs.epoll.max_user_watches
+    value: 785592
+  - name: fs.file-max
+    state: absent
+  - name: kernel.threads-max
+    state: absent
+```
+This will add/modify the parameter `fs.epoll.max_user_watches` in the managed
+object (in this case the managed object is the OS `sysctl` settings), and
+remove the the parameters `fs.file-max` and `kernel.threads-max` (in this
+case, this means the values will revert back to their system default values,
+usually upon system reboot).
+
+## Replacing an entire list of items with the given values with previous: replaced
+
+This is the case where you want to remove all of the existing values in the
+managed object and replace those values with the given values.  Use `previous:
+replaced` as one of the items in the list of values (preferably the first
+value in the list).  For example, using `kernel_settings_sysctl` parameters:
+```yaml
+kernel_settings_sysctl:
+  - previous: replaced
+  - name: fs.epoll.max_user_watches
+    value: 785592
+  - name: fs.file-max
+    value: 111111
+  - name: kernel.threads-max
+    value: 222222
+```
+This will remove any existing settings and replace them with the settings given.
+
+## Removing all of the settings with state: empty
+
+This is the case where you want to remove all of the existing values in the
+managed object.  For example, if you want to remove all of the
+`kernel_settings_sysctl` parameters:
+```yaml
+kernel_settings_sysctl:
+  state: empty
```

Comment:
It could be mentioned that this is exactly equivalent, but more readable, version of
```
kernel_settings_sysctl:
  - previous: replaced
```
as it specifies a list that is empty except the `previous: replaced` special value which drops the previous settings.

Reply:
addressed in https://github.com/linux-system-roles/linux-system-roles.github.io/pull/21/commits/4577d2c20cb2a14ed176486cb823442645d7088b